jobs/ads-incrementality-dap-collector/ads_incrementality_dap_collector/helpers.py

```diff
@@ -0,0 +1,331 @@
+                "--hpke-config",
+                hpke_config,
+                "--hpke-private-key",
+                hpke_private_key,
+            ],
+            capture_output=True,
+            text=True,
+            check=True,
+            timeout=PROCESS_TIMEOUT,
+        )
+        for line in result.stdout.splitlines():
+            if line.startswith("Aggregation result:"):
+                entries = parse_histogram(line[21:-1])
+                return entries
+    # Beware! Exceptions thrown by the subprocess reveal secrets.
+    # Log them and include traceback only for debugging in local dev.
+    except subprocess.CalledProcessError as e:
+        raise Exception(
+            f"Collection failed for {task_id}, {e.returncode}, stderr: {e.stderr}"
+        ) from None
+    except subprocess.TimeoutExpired as e:
+        raise Exception(
+            f"Collection timed out for {task_id}, {e.timeout}, stderr: {e.stderr}"
+        ) from None
+
+
+def collect_dap_results(
+    tasks_to_collect: dict[str, dict[int, IncrementalityBranchResultsRow]],
+    config: SimpleNamespace,
+    experiment_config: SimpleNamespace,
+):
+    tasks = list(dict.fromkeys(tasks_to_collect))
+    logging.info(f"Starting DAP collection for tasks: {tasks}.")
+    for task_id in tasks:
+        logging.info(f"Collecting DAP task: {task_id}")
+        results = tasks_to_collect[task_id]
+        # The task vector length and batch duration are specified per-experiment and
+        # stored with each branch. So it's okay to just use the first branch
+        # to populate these values for all the tasks here.
+        firstBranch = list(results.values())[0]
+        task_veclen = firstBranch.task_veclen
+        batch_start_epoch = int(
+            datetime.combine(firstBranch.batch_start, datetime.min.time()).timestamp()
+        )
+        batch_duration = firstBranch.batch_duration
+        collected = collect_dap_result(
+            task_id,
+            task_veclen,
+            batch_start_epoch,
+            batch_duration,
+            config.auth_token,
+            config.hpke_config,
+            config.hpke_private_key,
+        )
+        try:
+            for bucket in results.keys():
+                tasks_to_collect[task_id][bucket].value_count = collected[bucket]
+        except Exception as e:
+            raise Exception(
+                f"Failed to parse collected DAP results: {collected}"
+            ) from e
+        logging.info(f"Prepared final result rows: {tasks_to_collect[task_id]}")
+        logging.info(f"Finished collecting DAP task: {task_id}")
+    logging.info("Finished DAP collection for all tasks.")
+    return tasks_to_collect
+
+
+# TODO Trigger Airflow errors
+def correct_wraparound(num: int) -> int:
+    field_prime = 340282366920938462946865773367900766209
+    field_size = 128
+    cutoff = 2 ** (field_size - 1)
+    if num > cutoff:
+        logging.info(f"Corrected {num} to {num - field_prime} ")
```

Comment:
We should drop this log since it will output the DAP aggregated result.

Reply:
Good eye, thank you!